### PR TITLE
JCLOUDS-1116: Delegate endpoint should be preferred to the caller's one

### PR DESCRIPTION
**Pull request by nacx**
https://issues.apache.org/jira/browse/JCLOUDS-1116

When computing the endpoint for delegate apis, if both the caller and the callee define the endpoint, then the one in the callee should take precedence.

### CODE REVIEW

File: core/src/test/java/org/jclouds/rest/internal/RestAnnotationProcessorTest.java [hunk 1]
```diff
@@ -411,6 +418,46 @@ public HttpResponse invoke(HttpCommand command) {
...
+         @Override
+         public HttpResponse invoke(HttpCommand command) {
+            assertEquals(command.getCurrentRequest().getRequestLine(), "GET http://localhost:1111/client/1 HTTP/1.1");
+            return HttpResponse.builder().build();
+         }
+      });
+
+      try {
+         child.getInstance(Callee.class);
+         fail("Callee shouldn't be bound yet");
```
Comment (andrewgaul) [inline core/src/test/java/org/jclouds/rest/internal/RestAnnotationProcessorTest.java]: Might want to use assertj `failBecauseExceptionWasNotThrown`.

**Comment by andrewgaul** [comment]
:+1:

File: core/src/test/java/org/jclouds/rest/internal/RestAnnotationProcessorTest.java (hunk 1 above)
Comment (nacx) [inline core/src/test/java/org/jclouds/rest/internal/RestAnnotationProcessorTest.java]: Done!

File: core/src/main/java/org/jclouds/rest/internal/RestAnnotationProcessor.java [hunk 2]
```diff
@@ -198,14 +198,19 @@ public GeneratedHttpRequest apply(Invocation invocation) {
...
             logger.trace("using endpoint %s from invocation.getArgs() for %s", endpoint, invocation);
-      } else if (caller != null) {
-         endpoint = getEndpointFor(caller);
-         if (endpoint.isPresent())
-            logger.trace("using endpoint %s from caller %s for %s", endpoint, caller, invocation);
-         else
-            endpoint = findEndpoint(invocation);
       } else {
-         endpoint = findEndpoint(invocation);
+         endpoint = getEndpointFor(invocation);
```
Comment (zack-shoylev) [inline core/src/main/java/org/jclouds/rest/internal/RestAnnotationProcessor.java]: Document the new logic a little bit for future reference?

**Comment by nacx** [comment]
Squashed, rebased, added comment and merged to master as [040df11d](http://git-wip-us.apache.org/repos/asf/jclouds/commit/040df11d). Thanks for the review!

**Pull request closed without merging**